src/services/jsTyping.ts

```diff
@@ -0,0 +1,293 @@
+         * @param nodeModulesPath is the path to the "node_modules" folder
+         */
+        function getTypingNamesFromNodeModuleFolder(nodeModulesPath: string, filesToWatch: string[]) {
+            // Todo: add support for ModuleResolutionHost too
+            if (!host.directoryExists(nodeModulesPath)) {
+                return;
+            }
+
+            const typingNames: string[] = [];
+            const packageJsonFiles =
+                filter(
+                    host.readDirectory(nodeModulesPath, /*extension*/ undefined, /*exclude*/ undefined, /*depth*/ 2),
+                    f => ts.getBaseFileName(f) === "package.json");
+
+            for (const packageJsonFile of packageJsonFiles) {
+                const packageJsonDict = tryParseJson(packageJsonFile, host);
+                if (!packageJsonDict) { continue; }
+
+                filesToWatch.push(packageJsonFile);
+
+                // npm 3 has the package.json contains a "_requiredBy" field
+                // we should include all the top level module names for npm 2, and only module names whose
+                // "_requiredBy" field starts with "#" or equals "/" for npm 3.
+                if (packageJsonDict._requiredBy &&
+                    filter(packageJsonDict._requiredBy, (r: string) => r[0] === "#" || r === "/").length === 0) {
+                    continue;
+                }
+
+                // If the package has its own d.ts typings, those will take precedence. Otherwise the package name will be used
+                // to download d.ts files from DefinitelyTyped
+                const packageName = packageJsonDict["name"];
+                if (hasProperty(packageJsonDict, "typings")) {
+                    const absPath = ts.getNormalizedAbsolutePath(packageJsonDict.typings, ts.getDirectoryPath(packageJsonFile));
+                    inferredTypings[packageName] = absPath;
+                }
+                else {
+                    typingNames.push(packageName);
+                }
+            }
+            mergeTypings(typingNames);
+        }
+
+        function getTypingNamesFromCompilerOptions(options: CompilerOptions) {
+            const typingNames: string[] = [];
+            if (!options) {
+                return;
+            }
+            mergeTypings(typingNames);
+        }
+    }
+
+    /**
+     * Keep a list of typings names that we know cannot be obtained at the moment (could be because
+     * of network issues or because the package doesn't hava a d.ts file in DefinitelyTyped), so
+     * that we won't try again next time within this session.
+     * @param newTypingNames The list of new typings that the host attempted to acquire
+     * @param cachePath The path to the tsd.json cache
+     * @param host The object providing I/O related operations.
+     */
+    export function updateNotFoundTypingNames(newTypingNames: string[], cachePath: string, host: TypingResolutionHost): void {

```

Comment:
I do not think this is the correct API. we should not be caching these. i think we need to limit what discovertypings does.. it should just return a list of typeing "names" that it deemed applicable to this project, and a list of "locations" that it used to get to these inferences.
we should also add another interface that infers typings from fileNames, or just do this on the host side.
the host can then cache that list, call tsd, or any other tool ti likes, for all of them or some of them.. it can cahce what changed and what was not found.